nengo_spinnaker/operators/value_sink.py

```diff
@@ -41,71 +55,138 @@ def make_vertices(self, model, n_steps):  # TODO remove n_steps
+        }
+
+        # Determine resources usage
+        resources = {
+            Cores: 1,
+            SDRAM: sizeof_regions_named(self.regions, self.region_arguments)
+        }
+        super(ValueSinkVertex, self).__init__(get_application("value_sink"),
+                                              resources)
+
+    def accepts_signal(self, signal_params, transmission_params):
+        """Choose whether to receive this signal or not."""
+        if isinstance(transmission_params, EnsembleTransmissionParameters):
+            # If the connection is from an ensemble only return true if the
+            # decoders contain non-zero values in the input dimensions we care
+            # about.
+            return np.any(transmission_params.decoders[self.input_slice, :])
+        elif isinstance(transmission_params,
+                        PassthroughNodeTransmissionParameters):
+            # If the connection is from a Node of some variety then only return
+            # true if the transform contains non-zero values in the rows which
+            # relate to the subspace we receive input in.
+            return np.any(transmission_params.transform[self.input_slice])
+
+        # We don't know how to interpret the transmission parameters
+        raise NotImplementedError
+
+    def load_to_machine(self, netlist):
+        # Get a block of memory for each of the regions
+        self.region_memory = \
+            regions.utils.create_app_ptr_and_region_files_named(
+                netlist.vertices_memory[self], self.regions,
+                self.region_arguments
+            )
+
+        # Write the regions into memory
+        for key in Regions:
+            # Get the arguments
+            args, kwargs = self.region_arguments[key]
+
+            # Get the region
+            self.regions[key].write_subregion_to_file(
+                self.region_memory[key], *args, **kwargs
+            )
+
+    def read_recording(self, n_steps):
+        """Read back the recorded values."""
+        # Grab the block of memory and seek to the start
+        mem = self.region_memory[Regions.recording]
+        mem.seek(0)
+
+        # Perform the read
+        return self.regions[Regions.recording].to_array(
+            mem, self.input_slice, n_steps
+        )
 
 
 class SystemRegion(regions.Region):
     """System region for a value sink."""
-    def __init__(self, timestep, size_in):
+    def __init__(self, timestep, input_slice):
         self.timestep = timestep
-        self.size_in = size_in
+        self.input_slice = input_slice
 
     def sizeof(self, *args):
-        return 8  # 2 words
+        return 12  # 3 words
 
     def write_subregion_to_file(self, fp, *args):
-        fp.write(struct.pack("<2I", self.timestep, self.size_in))
+        size_in = self.input_slice.stop - self.input_slice.start
+        fp.write(struct.pack("<3I", self.timestep,
+                             size_in, self.input_slice.start))
```

Comment:
Entirely understood!

On 4 April 2016 at 10:17, Andrew Mundy <notifications@github.com> wrote:

> In nengo_spinnaker/operators/value_sink.py
> <https://github.com/project-rig/nengo_spinnaker/pull/116#discussion_r58347821>
> :
>
> >
> >      def write_subregion_to_file(self, fp, *args):
> > -        fp.write(struct.pack("<2I", self.timestep, self.size_in))
> > +        size_in = self.input_slice.stop - self.input_slice.start
> > +        fp.write(struct.pack("<3I", self.timestep,
> > +                             size_in, self.input_slice.start))
>
> I think this would be a good idea, there's actually quite a lot of
> neatening that could be done but that may be a job for a later PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/project-rig/nengo_spinnaker/pull/116/files/9268ded001922c79f638cc57e8c45f28fae6c95b#r58347821>
>